### DISCUSSION TITLE
1104 - Mystique EVM and Protocol Upgrades

### DISCUSSION THREAD

**Discussion by realcodywburns** [ethereumclassic/ECIPs#440] [ECIPs]
<html>
<body>
<p style="box-sizing: border-box; margin-bottom: 16px; margin-top: 0px !important; color: rgb(173, 186, 199); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(34, 39, 46); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Enable the outstanding Ethereum Foundation<span> </span><em style="box-sizing: border-box;">London</em><span> </span>network protocol upgrades on<br style="box-sizing: border-box;">the Ethereum Classic network in a hard-fork code-named<span> </span><em style="box-sizing: border-box;">Mystique</em><span> </span>to enable<br style="box-sizing: border-box;">maximum compatibility across these networks.</p><h3 style="box-sizing: border-box; margin-bottom: 16px; margin-top: 24px; font-size: 1.25em; font-weight: 600; line-height: 1.25; color: rgb(173, 186, 199); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(34, 39, 46); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Abstract</h3><p style="box-sizing: border-box; margin-bottom: 16px; margin-top: 0px; color: rgb(173, 186, 199); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(34, 39, 46); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Add support for a subset of protocol-impacting changes introduced in the<br style="box-sizing: border-box;">Ethereum Foundation (ETH) network via the<span> </span><em style="box-sizing: border-box;">London</em><span> </span>hardforks. The proposed<br style="box-sizing: border-box;">changes for Ethereum Classic's<span> </span><em style="box-sizing: border-box;">London</em><span> </span>upgrade include:</p>

EIP | Eth PR | omit or include
-- | -- | --
3198 (BASEFEE opcode) | #22837 | Include
3529 (Alternative refund reduction) | #22733 | Include
3541 (Reject new contracts starting with the 0xEF byte) | #22809 | Include
1559 (Fee market change) | #22837 #22896 | Omit
3228 (💣 delay) | #22840 and #22870 | Omit

<h3 style="box-sizing: border-box; margin-bottom: 16px; margin-top: 24px; font-size: 1.25em; font-weight: 600; line-height: 1.25; color: rgb(173, 186, 199); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(34, 39, 46); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Specification</h3><p style="box-sizing: border-box; margin-bottom: 16px; margin-top: 0px; color: rgb(173, 186, 199); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(34, 39, 46); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Technical specifications for each EIP can be found at those documents<br style="box-sizing: border-box;">respectively:</p><ul style="box-sizing: border-box; margin-bottom: 0px !important; margin-top: 0px; padding-left: 2em; color: rgb(173, 186, 199); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(34, 39, 46); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box; margin-left: 0px;"><a href="https://eips.ethereum.org/EIPS/eip-3198" rel="nofollow" style="box-sizing: border-box; background-color: transparent; color: var(--color-text-link); text-decoration: none;">EIP-3198</a>: BASEFEE opcode</li><li style="box-sizing: border-box; margin-top: 0.25em; margin-left: 0px;"><a href="https://eips.ethereum.org/EIPS/eip-3529" rel="nofollow" style="box-sizing: border-box; background-color: transparent; color: var(--color-text-link); text-decoration: none;">EIP-3529</a>: Alternative refund reduction</li><li style="box-sizing: border-box; margin-top: 0.25em; margin-left: 0px;"><a href="https://eips.ethereum.org/EIPS/eip-3541" rel="nofollow" style="box-sizing: border-box; background-color: transparent; color: var(--color-text-link); text-decoration: none;">EIP-3541</a>: Reject new contracts starting with the 0xEF byte</li><li style="box-sizing: border-box; margin-top: 0.25em; margin-left: 0px;"><a href="https://eips.ethereum.org/EIPS/eip-1559" rel="nofollow" style="box-sizing: border-box; background-color: transparent; color: var(--color-text-link); text-decoration: none;">EIP-1559</a>: Fee market change</li><li style="box-sizing: border-box; margin-top: 0.25em; margin-left: 0px;"><a href="https://eips.ethereum.org/EIPS/eip-3228" rel="nofollow" style="box-sizing: border-box; background-color: transparent; color: var(--color-text-link); text-decoration: none;">EIP-3228</a>:<span> </span><g-emoji class="g-emoji" alias="bomb" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4a3.png" style="box-sizing: border-box; font-family: &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; font-size: 1em; font-weight: 400; line-height: 1; vertical-align: -0.075em; font-style: normal !important;">💣</g-emoji> delay</li></ul>
</body>
</html>

**Comment by stevanlohja** [comment]
📅  Save the date, Aug 20, 2021, Mystique is part of the meeting agenda. https://github.com/ethereumclassic/ECIPs/issues/434#issue-965694466

**Comment by tvinhas** [comment]
Can we file a new EIP to ban Donald McIntyre from ETC?

**Reply by OmniEdge** [reply]
Good poke :-) Of course anyone can file a new EIP. Banning historically only happens in EIP. On ECIP it wont find rough consensus. Keeping that culture in ETC alive is what makes that fine but crucial difference.

**Reply by DonaldMcIntyre** [reply]
> Can we file a new EIP to ban Donald McIntyre from ETC?

<img width="1680" alt="chad do it large" src="https://user-images.githubusercontent.com/79114169/130327774-d18dfc5c-9965-48d7-8d87-ce4ca3f9ee65.png">

**Comment by realcodywburns** [comment]
further context around eip-3541: 

 The evm object format [eip-3540](https://eips.ethereum.org/EIPS/eip-3540) is the intended future state of the evm for blockchain in the future.

The EOF format requires having unique prefix bytes not previously deployed to the blockchain. To freeze the search space, 
EIP 3541 (https://eips.ethereum.org/EIPS/eip-3541) has been activated in the Ethereum London upgrade this August.
If we consider using EOF in future, and be wish to compatible with the EOF prefix, we should activate EIP-3541 as soon as possible.

**Comment by iquidus** [comment]
I propose aiming for mainnet activation at block 14,308,403, this will land mid January after the holiday season.

**Reply by bobsummerwill** [reply]
That makes sense to me.  Would you agree, @diega?

**Reply by diega** [reply]
I totally agree

**Comment by realcodywburns** [comment]
Are the clients ready for mordor activation next week?

**Reply by diega** [reply]
Tomorrow we'll perform an internal test with @ziogaschr to define

**Reply by diega** [reply]
update: we attempted a private fork activation and it failed, probably regarding some issue on the mining module

**Comment by diega** [comment]
After our testing session we may need to revisit some of the decisions made for **Mystique** around adding [EIP-3198](https://eips.ethereum.org/EIPS/eip-3198).

EIP-3198 **requires** [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559) and, as such, its behavior when there is no _baseFee_ block header is undetermined. E.g. Besu [implements it](https://github.com/hyperledger/besu/blob/21.10.2/evm/src/main/java/org/hyperledger/besu/evm/operation/BaseFeeOperation.java#L38) halting the EVM with an `INVALID_OPERATION` but at the same time consumes `2` units of gas.

I think our options can be summarized as follow:
1. not implement EIP-1559, not implement EIP-3198
2. not implement EIP-1559 and make `BASEFEE` to return `0`
3. partially implement EIP-1559 adding the _baseFee_ header always equal to `0` and implement EIP-3198  

I would personally advocate for **1.** because if ETC is not implementing EIP-1559, any behavior we decide for EIP-3198 won't keep compatibility

In any case, I would propose to postpone every date mentioned in [ECIP-1104](https://github.com/ethereumclassic/ECIPs/pull/452) by one or two weeks.

cc: @ziogaschr

**Comment by bobsummerwill** [comment]
I agree that (1) is probably the best choice.  Better to omit both EIPs.

1559-style transactions aren't possible in ETC, so use of BASEFEE does not make sense for ETC either.

I am also supportive of a deferral of Mordor/Kotti activation for Mystique while we work through this specification discussion and until we have working client software for that specification.

**Comment by ziogaschr** [comment]
Based on the partially implementation of (3), I think I prefer it, as if it's not that much code, then it worths keeping ETH backward compatibility as long as possible. On the other hand, I don't think many contracts are using the `BASEFEE` opcode already.

**Comment by q9f** [comment]
I'm leaning towards 3 and agree to postpone the activation a bit. It seems we need more insights first.

**Comment by DonaldMcIntyre** [comment]
Seems the choice that maximizes EVM compatibility in the present and going forward without changing the monetary policy (e.g. no changes in fee model nor fee burning) would be most appropriate.

**Comment by diega** [comment]
I'm not quite sure that we can say we are EIP-3198 compatible if we don't fully implement EIP-1559. I don't think the value returned would have any valid semantic in the terms of the spec of EIP-3198.
Regarding (3), I'm not convinced that changing the block format in order to hardcode a value is a good idea.

**Comment by realcodywburns** [comment]
Watched the dev call, all seems good. I think it is good to move forward with the fork even if it is only two seemingly minor changes. omitting eips is as important as including them sometimes.

**Comment by diega** [comment]
FYI, I just updated #452 with the agreements of the call, it's up to review

**Comment by stevanlohja** [comment]
Should the proposal be updated to `Accepted` status?

**Reply by diega** [reply]
You're correct. This is the corresponding [PR](https://github.com/ethereumclassic/ECIPs/pull/459)
